src/Device/Driver/AR62xx.cpp

```diff
@@ -0,0 +1,496 @@
+{
+  /** check if we received data */
+  if (length == 0)
+    return false;
+
+  /** try to parse data, if it can be parsed it will return true*/
+  bool data_is_ok = ParseReceivedData((const uint8_t *)_data, length);
+
+  /** send nmea info that we are receiving data */
+  if (data_is_ok)
+    info.alive.Update(info.clock);
+
+  return data_is_ok;
+}
+
+bool AR62xxDevice::Send(const uint8_t *msg, size_t msg_size, OperationEnvironment &env)
+{
+  /** Number of tries to send a message will be decreased on every retry */
+  unsigned retries = 5;
+
+  /** flag if message has been sent */
+  bool message_sent = false;
+
+  /** loop until retries are reached of message has been sent */
+  do
+  {
+    /** try to send message */
+    message_sent = port.FullWrite(msg, msg_size, env, std::chrono::milliseconds(250));
+
+    retries--;
+  } while (retries > 0 && !message_sent);
+
+  return message_sent;
+}
+
+RadioFrequency AR62xxDevice::AR62xxIdToFrequency(const uint16_t frequency_id)
+{
+  /** get frequency-part out of frequency-id */
+  unsigned frequency_encoded = frequency_id & FREQUENCY_BITMASK;
+
+  /** get channel-part out of frequency-id */
+  unsigned channel_encoded = frequency_id & CHANNEL_BITMASK;
+
+  /** calculate the frequency */
+  unsigned frequency = MIN_FREQUENCY + frequency_encoded * FREQUENCY_RANGE / MAX_FREQUENCY_ID;
+
+  /** add channel to frequency */
+  if (channel_encoded <= 3)
+    frequency += channel_encoded * 5;
+  else if (channel_encoded <= 7)
+    frequency += channel_encoded * 5 + 5;
+  else if (channel_encoded <= 11)
+    frequency += channel_encoded * 5 + 10;
+  else if (channel_encoded <= 15)
+    frequency += channel_encoded * 5 + 15;
+
+  /** generate RadioFrequency object */
+  RadioFrequency radio_frequency = RadioFrequency();
+  radio_frequency.SetKiloHertz(frequency);
+
+  return radio_frequency;
+}
+
+uint16_t AR62xxDevice::FrequencyToAR62xxId(const RadioFrequency freq)
+{
+  /** get frequency in kiloherz */
+  unsigned frequency = freq.GetKiloHertz();
+
+  /** check if frequency is not in range */
+  if (frequency < MIN_FREQUENCY || frequency > MAX_FREQUENCY)
+    return 0;
+
+  /** get offset of frequency to the lowest frequency which is possible */
+  unsigned frequency_offset = frequency - MIN_FREQUENCY;
+
+  /** generate frequency-id */
+  uint16_t frequency_id = std::round(frequency_offset * MAX_FREQUENCY_ID / FREQUENCY_RANGE);
+
+  /** append frequency bitmask as channel has to be added separately */
+  frequency_id &= FREQUENCY_BITMASK;
+
+  /** get channel out of frequency */
+  uint8_t channel = frequency % 100;
+
+  /** encode channel and add to frequency */
+  if (channel <= 15)
+    frequency_id += channel / 5;
+  else if (channel <= 40)
+    frequency_id += channel / 5 - 1;
+  else if (channel <= 65)
+    frequency_id += channel / 5 - 2;
+  else if (channel <= 90)
+    frequency_id += channel / 5 - 3;
+
+  return (frequency_id);
+}
+
+static uint16_t CRCBitwise(uint8_t *data, size_t len)
```

Comment:
Does this function implement a standard CRC algorithm or is this a proprietary CRC algorithm?

Reply:
good question ... i'll try to find out

Reply:
..... and why does it require a writable buffer?

Reply:
added `const`

Reply:
this is a 16-bit binary polynomial represented by number 0x8005